packages/kit/src/exports/public.d.ts

```diff
@@ -1937,6 +1937,46 @@ export interface RemoteFormIssue {
 	message: string;
 }
 
+/**
+ * Recursively maps an input type to a structure where each field can create a validation issue.
+ * This mirrors the runtime behavior of the `invalid` proxy passed to form handlers.
+ */
+type InvalidField<T> =
+	WillRecurseIndefinitely<T> extends true
+		? Record<string | number, any>
+		: NonNullable<T> extends string | number | boolean | File
+			? (message: string) => StandardSchemaV1.Issue
+			: NonNullable<T> extends Array<infer U>
+				? {
+						[K in number]: InvalidField<U>;
+					} & ((message: string) => StandardSchemaV1.Issue)
+				: NonNullable<T> extends RemoteFormInput
+					? {
+							[K in keyof T]-?: InvalidField<T[K]>;
+						} & ((message: string) => StandardSchemaV1.Issue)
+					: Record<string, never>;
+
+/**
+ * A function and proxy object used to imperatively create validation errors in form handlers.
+ *
+ * Call `invalid([...])` with an array of issues to throw a validation error.
+ * Access properties to create field-specific issues: `invalid.fieldName('message')`.
+ * The type structure mirrors the input data structure for type-safe field access.
+ *
+ * @example
+ * ```ts
+ * invalid([
+ *   invalid.username('Username is taken'),
+ *   invalid.items[0].qty('Insufficient stock')
+ * ]);
+ * ```
+ */
+export type Invalid<Input = any> = ((issues: StandardSchemaV1.Issue[]) => never) &
+	InvalidField<Input> & {
+		/** Create an issue for the root of the form */
+		$: (message: string) => StandardSchemaV1.Issue;
```

Comment:
this is an interesting idea. Can't decide if I like it or not. Alternative idea: what if you could pass a string? Then the common case would be nice and easy:

```diff
try {
  await auth.api.signInEmail({
    body: { email, password }
  });
} catch (e) {
  if (e instanceof APIError) {
-   invalid([
-     invalid.$(e.message)
-   ]);
+   invalid([e.message]);
  }

  throw e;
}
```

Another thought, now that I'm actually playing with it: in the vast majority of programmatic validation cases, you'll be in a `catch` clause and there'll be a single error. So what if instead of `invalid(issues)` we did `invalid(...issues)`? For the common case it feels nicer I think.